File: zy73133/pipeline.py
from __future__ import annotations

import csv
import hashlib
import json
import os
import traceback
from copy import deepcopy
from datetime import datetime
from typing import List, Dict, Any, Optional, Tuple

from models import (
    TideStationAnnotation, AnnotationVersion, RemarkDelta,
    ExportRecord, BatchProcessResult, AnnotationStatus, generate_id
)
from annotation_engine import (
    parse_buoy_log_line, build_buoy_record, create_annotation, CSV_HEADERS
)


import re as _re


def _parse_remark_corrections(remark_text: str) -> Dict[str, Any]:
    result: Dict[str, Any] = {
        "tide_offset_m": 0.0,
        "tide_override": None,
        "found": False,
    }

    offset_match = _re.search(r'(减去|减掉|扣除|减去|\+|\-|加|增加)\s*(-?\d+\.?\d*)\s*(m|米|公尺)', remark_text)
    if offset_match:
        op = offset_match.group(1)
        val = float(offset_match.group(2))
        if op in ('减去', '减掉', '扣除', '-'):
            result["tide_offset_m"] = -val
        else:
            result["tide_offset_m"] = val
        result["found"] = True

    override_match = _re.search(r'(实际潮位|修正为|订正为|实际为)\s*[:：]?\s*(-?\d+\.?\d*)\s*(m|米|公尺)?', remark_text)
    if override_match:
        result["tide_override"] = float(override_match.group(2))
        result["found"] = True

    return result


def _apply_corrections_to_record(record, corrections: Dict[str, Any]):
    from copy import deepcopy
    new_record = deepcopy(record)
    if new_record.tide is None:
        return new_record

    if corrections.get("tide_override") is not None:
        old_val = new_record.tide.value_meters
        new_val = corrections["tide_override"]
        new_record.tide.value_meters = round(new_val, 6)
        new_record.tide.normalize_notes.append(
            f"[备注修正] 人工订正潮位由 {old_val:.4f}m 改为 {new_val:.4f}m"
        )
    elif corrections.get("tide_offset_m", 0) != 0:
        offset = corrections["tide_offset_m"]
        old_val = new_record.tide.value_meters
        new_val = round(old_val + offset, 6)
        new_record.tide.value_meters = new_val
        sign = "+" if offset >= 0 else ""
        new_record.tide.normalize_notes.append(
            f"[备注修正] 潮位人工修正: {old_val:.4f}m {sign}{offset:.4f}m = {new_val:.4f}m"
        )
    return new_record


def _compare_annotations(old: TideStationAnnotation, new: TideStationAnnotation) -> List[RemarkDelta]:
    deltas: List[RemarkDelta] = []

    fields_to_compare = [
        ("scene_annotation", "场景标注文本"),
        ("side_note", "侧边说明文本"),
        ("status", "状态"),
    ]

    for field, label in fields_to_compare:
        old_val = getattr(old, field)
        new_val = getattr(new, field)
        if old_val != new_val:
            deltas.append(RemarkDelta(
                field_changed=label,
                old_value=old_val,
                new_value=new_val,
                judgment_impact=f"{label}发生变更，影响输出一致性"
            ))

    old_row = old.csv_row
    new_row = new.csv_row
    csv_fields = [
        ("tide_level", "潮位等级判断"),
        ("tide_meters", "标准化潮位值"),
        ("remarks", "备注内容"),
        ("issues", "解析问题线索"),
    ]
    for field, label in csv_fields:
        if old_row.get(field) != new_row.get(field):
            if field == "tide_level":
                impact = f"潮位等级由【{old_row.get(field)}】变更为【{new_row.get(field)}】，影响站点可开发性判断"
            elif field == "remarks":
                impact = "新增/修改人工备注，影响判断依据，需重新评估"
            elif field == "tide_meters":
                impact = f"标准化潮位值由 {old_row.get(field)} m 变更为 {new_row.get(field)} m，影响潮位等级与能位判断"
            else:
                impact = f"{label}发生变更，影响问题追溯"
            deltas.append(RemarkDelta(
                field_changed=label,
                old_value=old_row.get(field, ""),
                new_value=new_row.get(field, ""),
                judgment_impact=impact
            ))

    return deltas


def apply_remark_and_version(
    annotation: TideStationAnnotation,
    remark_text: str,
    version_number: int
) -> Tuple[TideStationAnnotation, AnnotationVersion]:
    new_remarks = annotation.remarks + [remark_text]
    old_annotation = deepcopy(annotation)

    corrections = _parse_remark_corrections(remark_text)
    effective_record = annotation.buoy_record
    if corrections["found"]:
        effective_record = _apply_corrections_to_record(annotation.buoy_record, corrections)

    new_annotation = create_annotation(
        batch_id=annotation.batch_id,
        record=effective_record,
        remarks=new_remarks,
        status=AnnotationStatus.REPROCESSED
    )
    new_annotation.annotation_id = annotation.annotation_id
    new_annotation.created_at = annotation.created_at
    new_annotation.updated_at = datetime.now()
    new_annotation.raw_trace["remark_applied"] = remark_text
    new_annotation.raw_trace["remark_time"] = datetime.now().isoformat()
    new_annotation.raw_trace["previous_version_annotation_id"] = annotation.annotation_id
    if corrections["found"]:
        new_annotation.raw_trace["remark_corrections"] = corrections

    deltas = _compare_annotations(old_annotation, new_annotation)

    version = AnnotationVersion(
        version_id=generate_id('ver'),
        batch_id=annotation.batch_id,
        annotation_id=annotation.annotation_id,
        version_number=version_number,
        annotation=deepcopy(new_annotation),
        applied_remark=remark_text,
        deltas=deltas
    )

    return new_annotation, version


def format_delta_report(version: AnnotationVersion) -> str:
    if not version.deltas:
        return f"版本 {version.version_number}：备注未改变任何判断结果\n备注内容: {version.applied_remark}"

    lines = [
        f"=== 版本 {version.version_number} 变更报告 ===",
        f"备注内容: {version.applied_remark}",
        f"影响判断的变更 ({len(version.deltas)} 项):",
    ]
    for i, delta in enumerate(version.deltas, 1):
        lines.append(f"  {i}. {delta.field_changed}")
        lines.append(f"     旧值: {delta.old_value}")
        lines.append(f"     新值: {delta.new_value}")
        lines.append(f"     影响: {delta.judgment_impact}")
    return "\n".join(lines)


def safe_process_log_line(raw_text: str, source_file: str, line_number: int,
                          batch_id: str) -> Tuple[Optional[TideStationAnnotation], Optional[str]]:
    raw_log = None
    try:
        raw_log = parse_buoy_log_line(raw_text, source_file, line_number)
        record = build_buoy_record(raw_log)

        if not record.is_valid:
            annotation = create_annotation(
                batch_id=batch_id,
                record=record,
                status=AnnotationStatus.EXCEPTION
            )
            annotation.raw_trace["exception_context"] = {
                "raw_text": raw_text,
                "parse_errors": record.parse_errors,
                "raw_log_snapshot": {
                    "raw_latitude": raw_log.raw_latitude,
                    "raw_longitude": raw_log.raw_longitude,
                    "raw_tide_value": raw_log.raw_tide_value,
                    "raw_tide_unit": raw_log.raw_tide_unit,
                }
            }
            return annotation, None

        annotation = create_annotation(batch_id=batch_id, record=record)
        return annotation, None

    except Exception as e:
        error_trace = traceback.format_exc()
        annotation = TideStationAnnotation(
            annotation_id=generate_id('ann'),
            batch_id=batch_id,
            station_name="异常记录",
            buoy_record=record if 'record' in dir() else build_buoy_record(
                raw_log if raw_log else parse_buoy_log_line(raw_text, source_file, line_number)
            ),
            scene_annotation=f"[处理异常] {str(e)} | 原始日志: {raw_text}",
            side_note=f"处理异常: {str(e)}\n原始日志保留:\n{raw_text}\n来源: {source_file}:{line_number}\n堆栈:\n{error_trace}",
            csv_row={
                "station_name": "异常记录",
                "timestamp": "",
                "latitude": "",
                "longitude": "",
                "latitude_raw": raw_log.raw_latitude if raw_log else "",
                "longitude_raw": raw_log.raw_longitude if raw_log else "",
                "coordinate_format": "",
                "tide_meters": "",
                "tide_original": raw_log.raw_tide_value + " " + raw_log.raw_tide_unit if raw_log else "",
                "tide_unit_raw": raw_log.raw_tide_unit if raw_log else "",
                "tide_level": "",
                "remarks": "",
                "issues": f"EXCEPTION: {str(e)}; 原始线索已保留",
                "raw_text": raw_text,
                "source_ref": f"{source_file}:{line_number}",
                "record_id": "",
                "log_id": raw_log.log_id if raw_log else "",
            },
            status=AnnotationStatus.EXCEPTION,
            remarks=[f"处理异常: {str(e)}"],
            raw_trace={
                "exception": str(e),
                "traceback": error_trace,
                "raw_text": raw_text,
                "source_file": source_file,
                "line_number": line_number,
            }
        )
        return annotation, error_trace


def process_buoy_file(file_path: str, batch_id: str) -> List[TideStationAnnotation]:
    annotations: List[TideStationAnnotation] = []
    with open(file_path, 'r', encoding='utf-8') as f:
        for line_num, line in enumerate(f, 1):
            stripped = line.strip()
            if not stripped or stripped.startswith('#'):
                continue
            ann, _ = safe_process_log_line(stripped, file_path, line_num, batch_id)
            if ann:
                annotations.append(ann)
    return annotations


def _content_hash(text: str) -> str:
    return hashlib.sha256(text.encode('utf-8')).hexdigest()


def export_scene_annotations(annotations: List[TideStationAnnotation],
                             output_path: str, batch_id: str,
                             version_id: str) -> ExportRecord:
    lines = [ann.scene_annotation for ann in annotations]
    content = "\n".join(lines) + "\n"
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(content)
    return ExportRecord(
        export_id=generate_id('exp'),
        batch_id=batch_id,
        version_id=version_id,
        export_type="scene_annotations",
        export_path=output_path,
        content_hash=_content_hash(content),
        annotation_count=len(annotations)
    )


def export_side_notes(annotations: List[TideStationAnnotation],
                      output_path: str, batch_id: str,
                      version_id: str) -> ExportRecord:
    sections = []
    for ann in annotations:
        sections.append(f"===== {ann.station_name} [{ann.annotation_id}] =====")
        sections.append(ann.side_note)
        sections.append("")
    content = "\n".join(sections)
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(content)
    return ExportRecord(
        export_id=generate_id('exp'),
        batch_id=batch_id,
        version_id=version_id,
        export_type="side_notes",
        export_path=output_path,
        content_hash=_content_hash(content),
        annotation_count=len(annotations)
    )


def export_csv(annotations: List[TideStationAnnotation],
               output_path: str, batch_id: str,
               version_id: str) -> ExportRecord:
    with open(output_path, 'w', encoding='utf-8', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=CSV_HEADERS)
        writer.writeheader()
        for ann in annotations:
            row = {h: ann.csv_row.get(h, '') for h in CSV_HEADERS}
            writer.writerow(row)

    with open(output_path, 'r', encoding='utf-8') as f:
        content = f.read()
    return ExportRecord(
        export_id=generate_id('exp'),
        batch_id=batch_id,
        version_id=version_id,
        export_type="csv",
        export_path=output_path,
        content_hash=_content_hash(content),
        annotation_count=len(annotations)
    )


def export_version_trace(versions: List[AnnotationVersion],
                         output_path: str, batch_id: str) -> ExportRecord:
    reports = []
    for v in versions:
        reports.append(format_delta_report(v))
        reports.append("")
    content = "\n".join(reports)
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(content)
    return ExportRecord(
        export_id=generate_id('exp'),
        batch_id=batch_id,
        version_id=versions[-1].version_id if versions else '',
        export_type="version_trace",
        export_path=output_path,
        content_hash=_content_hash(content),
        annotation_count=len(versions)
    )


def verify_export_consistency(annotations: List[TideStationAnnotation],
                              csv_path: str, scene_path: str,
                              side_path: str) -> Dict[str, Any]:
    result: Dict[str, Any] = {
        "consistent": True,
        "mismatches": [],
        "count_check": {}
    }

    with open(scene_path, 'r', encoding='utf-8') as f:
        scene_lines = [l for l in f.read().splitlines() if l.strip()]
    if len(scene_lines) != len(annotations):
        result["consistent"] = False
        result["mismatches"].append(
            f"场景标注行数({len(scene_lines)})与标注数量({len(annotations)})不一致"
        )

    for i, ann in enumerate(annotations):
        if i < len(scene_lines) and scene_lines[i] != ann.scene_annotation:
            result["consistent"] = False
            result["mismatches"].append(f"第{i+1}条场景标注与内存标注不一致")

    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        csv_rows = list(reader)
    if len(csv_rows) != len(annotations):
        result["consistent"] = False
        result["mismatches"].append(
            f"CSV行数({len(csv_rows)})与标注数量({len(annotations)})不一致"
        )

    for i, ann in enumerate(annotations):
        if i < len(csv_rows):
            for key in CSV_HEADERS:
                if csv_rows[i].get(key, '') != ann.csv_row.get(key, ''):
                    result["consistent"] = False
                    result["mismatches"].append(
                        f"第{i+1}条CSV字段 {key} 不一致"
                    )
                    break

    result["count_check"] = {
        "annotations": len(annotations),
        "scene_lines": len(scene_lines),
        "csv_rows": len(csv_rows),
    }
    return result


def run_full_pipeline(source_files: List[str], output_dir: str,
                      extra_remarks: Optional[Dict[str, str]] = None
                      ) -> BatchProcessResult:
    extra_remarks = extra_remarks or {}
    os.makedirs(output_dir, exist_ok=True)

    batch_id = generate_id('batch')
    all_annotations: List[TideStationAnnotation] = []
    all_versions: List[AnnotationVersion] = []
    all_exports: List[ExportRecord] = []
    has_exceptions = False

    for src in source_files:
        anns = process_buoy_file(src, batch_id)
        all_annotations.extend(anns)

    for ann in all_annotations:
        if ann.status == AnnotationStatus.EXCEPTION:
            has_exceptions = True
        v0 = AnnotationVersion(
            version_id=generate_id('ver'),
            batch_id=batch_id,
            annotation_id=ann.annotation_id,
            version_number=1,
            annotation=deepcopy(ann),
            applied_remark=None,
            deltas=[]
        )
        all_versions.append(v0)

    v1_annotations = list(all_annotations)
    version_1_id = all_versions[-1].version_id if all_versions else generate_id('ver')

    scene_v1_path = os.path.join(output_dir, f"v1_scene_annotations_{batch_id}.txt")
    side_v1_path = os.path.join(output_dir, f"v1_side_notes_{batch_id}.txt")
    csv_v1_path = os.path.join(output_dir, f"v1_details_{batch_id}.csv")

    all_exports.append(export_scene_annotations(v1_annotations, scene_v1_path, batch_id, version_1_id))
    all_exports.append(export_side_notes(v1_annotations, side_v1_path, batch_id, version_1_id))
    all_exports.append(export_csv(v1_annotations, csv_v1_path, batch_id, version_1_id))

    if extra_remarks:
        reprocessed: List[TideStationAnnotation] = []
        for ann in all_annotations:
            remark = extra_remarks.get(ann.buoy_record.raw_log.log_id) or extra_remarks.get(ann.station_name)
            if remark:
                new_ann, new_version = apply_remark_and_version(ann, remark, 2)
                reprocessed.append(new_ann)
                all_versions.append(new_version)
            else:
                reprocessed.append(ann)

        version_2_id = all_versions[-1].version_id
        scene_v2_path = os.path.join(output_dir, f"v2_scene_annotations_{batch_id}.txt")
        side_v2_path = os.path.join(output_dir, f"v2_side_notes_{batch_id}.txt")
        csv_v2_path = os.path.join(output_dir, f"v2_details_{batch_id}.csv")
        trace_path = os.path.join(output_dir, f"version_trace_{batch_id}.txt")

        all_exports.append(export_scene_annotations(reprocessed, scene_v2_path, batch_id, version_2_id))
        all_exports.append(export_side_notes(reprocessed, side_v2_path, batch_id, version_2_id))
        all_exports.append(export_csv(reprocessed, csv_v2_path, batch_id, version_2_id))
        all_exports.append(export_version_trace(
            [v for v in all_versions if v.version_number == 2], trace_path, batch_id
        ))

        consistency_v2 = verify_export_consistency(reprocessed, csv_v2_path, scene_v2_path, side_v2_path)
        if not consistency_v2["consistent"]:
            has_exceptions = True

        all_annotations = reprocessed

    consistency_v1 = verify_export_consistency(
        [v.annotation for v in all_versions if v.version_number == 1],
        csv_v1_path, scene_v1_path, side_v1_path
    )
    if not consistency_v1["consistent"]:
        has_exceptions = True

    return BatchProcessResult(
        batch_id=batch_id,
        source_files=source_files,
        annotations=all_annotations,
        versions=all_versions,
        exports=all_exports,
        has_exceptions=has_exceptions
    )

File: zy73133/src/engine/pipeline.ts
import type {
  TideStationAnnotation,
  AnnotationVersion,
  RemarkDelta,
  RemarkCorrection,
  ConsistencyResult,
  BatchProcessResult,
  BuoyRecord,
  CsvRow,
} from './types';
import { generateId, CSV_HEADERS } from './types';
import { parseBuoyLogLine, buildBuoyRecord, createAnnotation } from './annotationEngine';

export function parseRemarkCorrections(remarkText: string): RemarkCorrection {
  const result: RemarkCorrection = { tideOffsetM: 0, tideOverride: null, found: false, description: '无可识别修正规则' };

  const overrideMatch = remarkText.match(/(?:实际潮位|修正为|订正为|实际为)\s*[:：]?\s*(-?\d+\.?\d*)\s*(m|米|公尺)?/);
  if (overrideMatch) {
    result.tideOverride = parseFloat(overrideMatch[1]);
    result.found = true;
    result.description = `识别订正规则：实际潮位修正为 ${overrideMatch[1]} m`;
    return result;
  }

  const offsetMatch = remarkText.match(/(减去|减掉|扣除|加|增加)\s*(-?\d+\.?\d*)\s*(m|米|公尺)/);
  if (offsetMatch) {
    const op = offsetMatch[1];
    const val = parseFloat(offsetMatch[2]);
    const sign = op === '减去' || op === '减掉' || op === '扣除' ? -1 : 1;
    result.tideOffsetM = sign * val;
    result.found = true;
    result.description = `识别修正规则：潮位${op} ${val} m（偏移 ${sign > 0 ? '+' : ''}${result.tideOffsetM} m）`;
    return result;
  }

  return result;
}

function applyCorrectionsToRecord(record: BuoyRecord, corrections: RemarkCorrection): BuoyRecord {
  const newRecord: BuoyRecord = {
    ...record,
    coordinates: record.coordinates ? { ...record.coordinates } : null,
    tide: record.tide ? { ...record.tide, normalizeNotes: [...record.tide.normalizeNotes] } : null,
    parseErrors: [...record.parseErrors],
    rawLog: { ...record.rawLog },
  };
  if (!newRecord.tide || newRecord.tide.valueMeters == null) return newRecord;

  if (corrections.tideOverride != null) {
    const oldVal = newRecord.tide.valueMeters;
    const newVal = corrections.tideOverride;
    newRecord.tide.valueMeters = Math.round(newVal * 1e6) / 1e6;
    newRecord.tide.normalizeNotes.push(`[备注修正] 人工订正潮位由 ${oldVal.toFixed(4)}m 改为 ${newVal.toFixed(4)}m`);
  } else if (corrections.tideOffsetM !== 0) {
    const offset = corrections.tideOffsetM;
    const oldVal = newRecord.tide.valueMeters;
    const newVal = Math.round((oldVal + offset) * 1e6) / 1e6;
    newRecord.tide.valueMeters = newVal;
    const sign = offset >= 0 ? '+' : '';
    newRecord.tide.normalizeNotes.push(
      `[备注修正] 潮位人工修正: ${oldVal.toFixed(4)}m ${sign}${offset.toFixed(4)}m = ${newVal.toFixed(4)}m`,
    );
  }
  return newRecord;
}

function compareAnnotations(oldAnn: TideStationAnnotation, newAnn: TideStationAnnotation): RemarkDelta[] {
  const deltas: RemarkDelta[] = [];

  const fields: { key: 'sceneAnnotation' | 'sideNote' | 'status'; label: string }[] = [
    { key: 'sceneAnnotation', label: '场景标注文本' },
    { key: 'sideNote', label: '侧边说明文本' },
    { key: 'status', label: '状态' },
  ];
  for (const { key, label } of fields) {
    const oldVal = String(oldAnn[key]);
    const newVal = String(newAnn[key]);
    if (oldVal !== newVal) {
      deltas.push({
        fieldChanged: label,
        oldValue: oldVal,
        newValue: newVal,
        judgmentImpact: `${label}发生变更，影响输出一致性`,
      });
    }
  }

  const csvFields: { key: keyof CsvRow; label: string }[] = [
    { key: 'tide_level', label: '潮位等级判断' },
    { key: 'tide_meters', label: '标准化潮位值' },
    { key: 'remarks', label: '备注内容' },
    { key: 'issues', label: '解析问题线索' },
  ];
  for (const { key, label } of csvFields) {
    const oldV = oldAnn.csvRow[key] ?? '';
    const newV = newAnn.csvRow[key] ?? '';
    if (oldV !== newV) {
      let impact: string;
      if (key === 'tide_level') {
        impact = `潮位等级由【${oldV}】变更为【${newV}】，影响站点可开发性判断`;
      } else if (key === 'remarks') {
        impact = '新增/修改人工备注，影响判断依据，需重新评估';
      } else if (key === 'tide_meters') {
        impact = `标准化潮位值由 ${oldV} m 变更为 ${newV} m，影响潮位等级与能位判断`;
      } else {
        impact = `${label}发生变更，影响问题追溯`;
      }
      deltas.push({ fieldChanged: label, oldValue: oldV, newValue: newV, judgmentImpact: impact });
    }
  }

  return deltas;
}

export function applyRemarkAndVersion(
  annotation: TideStationAnnotation,
  remarkText: string,
  versionNumber: number,
): { annotation: TideStationAnnotation; version: AnnotationVersion } {
  const newRemarks = [...annotation.remarks, remarkText];
  const oldAnnotation: TideStationAnnotation = {
    ...annotation,
    buoyRecord: {
      ...annotation.buoyRecord,
      coordinates: annotation.buoyRecord.coordinates ? { ...annotation.buoyRecord.coordinates } : null,
      tide: annotation.buoyRecord.tide
        ? { ...annotation.buoyRecord.tide, normalizeNotes: [...annotation.buoyRecord.tide.normalizeNotes] }
        : null,
    },
  };

  const corrections = parseRemarkCorrections(remarkText);
  let effectiveRecord = annotation.buoyRecord;
  if (corrections.found) {
    effectiveRecord = applyCorrectionsToRecord(annotation.buoyRecord, corrections);
  }

  const newAnnotation = createAnnotation(annotation.batchId, effectiveRecord, newRemarks, 'reprocessed');
  newAnnotation.annotationId = annotation.annotationId;
  newAnnotation.createdAt = annotation.createdAt;
  newAnnotation.updatedAt = new Date().toISOString();
  newAnnotation.rawTrace.remarkApplied = remarkText;
  newAnnotation.rawTrace.remarkTime = newAnnotation.updatedAt;
  newAnnotation.rawTrace.previousVersionAnnotationId = annotation.annotationId;
  if (corrections.found) newAnnotation.rawTrace.remarkCorrections = corrections;

  const deltas = compareAnnotations(oldAnnotation, newAnnotation);

  const version: AnnotationVersion = {
    versionId: generateId('ver'),
    batchId: annotation.batchId,
    annotationId: annotation.annotationId,
    versionNumber,
    annotation: {
      ...newAnnotation,
      buoyRecord: {
        ...newAnnotation.buoyRecord,
        coordinates: newAnnotation.buoyRecord.coordinates ? { ...newAnnotation.buoyRecord.coordinates } : null,
        tide: newAnnotation.buoyRecord.tide
          ? { ...newAnnotation.buoyRecord.tide, normalizeNotes: [...newAnnotation.buoyRecord.tide.normalizeNotes] }
          : null,
      },
    },
    appliedRemark: remarkText,
    deltas,
    createdAt: newAnnotation.updatedAt,
  };

  return { annotation: newAnnotation, version };
}

export function formatDeltaReport(version: AnnotationVersion): string {
  if (!version.deltas.length) {
    return `版本 ${version.versionNumber}：备注未改变任何判断结果\n备注内容: ${version.appliedRemark}`;
  }
  const lines = [
    `=== 版本 ${version.versionNumber} 变更报告 ===`,
    `备注内容: ${version.appliedRemark}`,
    `影响判断的变更 (${version.deltas.length} 项):`,
  ];
  version.deltas.forEach((d, i) => {
    lines.push(`  ${i + 1}. ${d.fieldChanged}`);
    lines.push(`     旧值: ${d.oldValue}`);
    lines.push(`     新值: ${d.newValue}`);
    lines.push(`     影响: ${d.judgmentImpact}`);
  });
  return lines.join('\n');
}

export function safeProcessLogLine(
  rawText: string,
  sourceFile: string,
  lineNumber: number,
  batchId: string,
): { annotation: TideStationAnnotation; error: string | null } {
  try {
    const rawLog = parseBuoyLogLine(rawText, sourceFile, lineNumber);
    const record = buildBuoyRecord(rawLog);

    if (!record.isValid) {
      const annotation = createAnnotation(batchId, record, [], 'exception');
      annotation.rawTrace.exceptionContext = {
        rawText,
        parseErrors: record.parseErrors,
        rawLogSnapshot: {
          rawLatitude: rawLog.rawLatitude,
          rawLongitude: rawLog.rawLongitude,
          rawTideValue: rawLog.rawTideValue,
          rawTideUnit: rawLog.rawTideUnit,
        },
      };
      return { annotation, error: null };
    }

    const annotation = createAnnotation(batchId, record);
    return { annotation, error: null };
  } catch (e) {
    const errorTrace = String(e?.stack ?? e);
    const rawLog = parseBuoyLogLine(rawText, sourceFile, lineNumber);
    const record = buildBuoyRecord(rawLog);
    const annotation = createAnnotation(batchId, record, [], 'exception');
    annotation.stationName = '异常记录';
    annotation.sceneAnnotation = `[处理异常] ${String(e)} | 原始日志: ${rawText}`;
    annotation.sideNote = `处理异常: ${String(e)}\n原始日志保留:\n${rawText}\n来源: ${sourceFile}:${lineNumber}\n堆栈:\n${errorTrace}`;
    annotation.csvRow = {
      ...annotation.csvRow,
      station_name: '异常记录',
      issues: `EXCEPTION: ${String(e)}; 原始线索已保留`,
      raw_text: rawText,
      source_ref: `${sourceFile}:${lineNumber}`,
    };
    annotation.rawTrace = {
      exception: String(e),
      traceback: errorTrace,
      rawText,
      sourceFile,
      lineNumber,
    };
    return { annotation, error: errorTrace };
  }
}

export function processBuoyText(
  text: string,
  sourceFile: string,
  batchId: string,
): TideStationAnnotation[] {
  const annotations: TideStationAnnotation[] = [];
  const lines = text.split(/\r?\n/);
  lines.forEach((line, idx) => {
    const stripped = line.trim();
    if (!stripped || stripped.startsWith('#')) return;
    const { annotation } = safeProcessLogLine(stripped, sourceFile, idx + 1, batchId);
    if (annotation) annotations.push(annotation);
  });
  return annotations;
}

export function runPipeline(logText: string, sourceName = 'pasted_logs.txt'): BatchProcessResult {
  const batchId = generateId('batch');
  const annotations = processBuoyText(logText, sourceName, batchId);

  const versions: AnnotationVersion[] = annotations.map((ann) => ({
    versionId: generateId('ver'),
    batchId,
    annotationId: ann.annotationId,
    versionNumber: 1,
    annotation: ann,
    appliedRemark: null,
    deltas: [],
    createdAt: ann.createdAt,
  }));

  const hasExceptions = annotations.some((a) => a.status === 'exception');

  return {
    batchId,
    annotations,
    versions,
    hasExceptions,
    currentVersion: 1,
  };
}

export function reprocessBatchWithRemark(
  batch: BatchProcessResult,
  annotationId: string,
  remarkText: string,
): { batch: BatchProcessResult; version: AnnotationVersion | null } {
  const nextVersion = batch.currentVersion + 1;
  let newVersion: AnnotationVersion | null = null;

  const newAnnotations = batch.annotations.map((ann) => {
    if (ann.annotationId !== annotationId) return ann;
    const { annotation: newAnn, version } = applyRemarkAndVersion(ann, remarkText, nextVersion);
    newVersion = version;
    return newAnn;
  });

  const newVersions = [...batch.versions];
  if (newVersion) newVersions.push(newVersion);

  return {
    batch: {
      ...batch,
      annotations: newAnnotations,
      versions: newVersions,
      currentVersion: nextVersion,
      hasExceptions: newAnnotations.some((a) => a.status === 'exception' || a.status === 'pending_review'),
    },
    version: newVersion,
  };
}

export function generateCsvString(annotations: TideStationAnnotation[]): string {
  const header = CSV_HEADERS.join(',');
  const rows = annotations.map((ann) =>
    CSV_HEADERS.map((h) => {
      const val = ann.csvRow[h] ?? '';
      const escaped = val.includes(',') || val.includes('"') || val.includes('\n')
        ? `"${val.replace(/"/g, '""')}"`
        : val;
      return escaped;
    }).join(','),
  );
  return [header, ...rows].join('\n');
}

export function verifyConsistency(annotations: TideStationAnnotation[]): ConsistencyResult {
  const mismatches: string[] = [];
  for (const ann of annotations) {
    if (!ann.csvRow.raw_text) mismatches.push(`${ann.stationName}: CSV 缺少 raw_text`);
    if (!ann.csvRow.source_ref) mismatches.push(`${ann.stationName}: CSV 缺少 source_ref`);
    if (!ann.sceneAnnotation.includes(ann.csvRow.tide_meters) && !ann.csvRow.tide_meters.includes('PARSE_FAILED')) {
      mismatches.push(`${ann.stationName}: 场景标注与 CSV 潮位值不一致`);
    }
    if (ann.status === 'exception' && ann.csvRow.latitude === '0.000000') {
      mismatches.push(`${ann.stationName}: 异常记录坐标被伪装为 0.000000`);
    }
    if (ann.status === 'exception' && !ann.csvRow.issues) {
      mismatches.push(`${ann.stationName}: 异常记录缺少失败原因`);
    }
  }
  return {
    consistent: mismatches.length === 0,
    mismatches,
    countCheck: {
      annotations: annotations.length,
      sceneLines: annotations.length,
      csvRows: annotations.length,
    },
  };
}

export function downloadCsv(csvString: string, filename: string): void {
  const blob = new Blob([`\uFEFF${csvString}`], { type: 'text/csv;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}
